assets/js/leaves.js

```diff
@@ -42,28 +42,42 @@ class Leaf {
   // constant for selection classname
   static selectedClass = "select";
 
-  static deselectAll() {
-    // deselect any leaf or leaf label that is currently highlighted
-    let selected = document.getElementsByClassName(Leaf.selectedClass);
-    // convert to array rather than iterating, since htmlcollection is live
-    // and changes as updated
-    Array.from(selected).forEach((item) => {
-      item.classList.remove(Leaf.selectedClass);
-    });
-
-    // change url location hash to indicate no leaf is selected
-    // needs to be non-empty to avoid page reload
-    window.location.replace("#");
-    history.replaceState(null, "", window.location.hash);
+  static setCurrentLeaf(event) {
+    // Are we removing the leaf?
+    if (event == undefined) {
+      // remove hash
+      let urlNoHash = window.location.pathname + window.location.search;
+      history.replaceState(null, "", urlNoHash);
+    } else {
+      // get the actual leaf target from DOM
+      let target = Leaf.getLeafTarget(event.target);
+      // get leaf ID
+      let leafID = target.dataset.id;
+      // update URL to reflect the currently selected leaf;
+      // replace the location & state to avoid polluting browser history
+      window.location.replace(`#${leafID}`);
+      history.replaceState(null, "", `#${leafID}`);
+    }
+    // regardless, update selection
+    Leaf.updateSelection();
   }
 
-  static selectByTag(tag) {
-    // select all leaves with the specified tag
-    Leaf.deselectAll();
-    let leaves = document.getElementsByClassName(tag);
-    for (let item of leaves) {
-      item.classList.add(Leaf.selectedClass);
+  static setCurrentTag(tag) {
+    // if no tag set, simply replace URL without tag param
+    if (tag == undefined) {
+      history.replaceState(
+        null,
+        "",
+        window.location.pathname + window.location.hash
+      );
+    } else {
+      // set URL to URL of self, with tag updated
+      let url = new URL(window.location.href);
+      url.searchParams.set("tag", tag);
```

Comment:
Here also, I think we can consolidate and use consistent syntax - use the `URL` object either way, and just `set` or `delete` the tag value depending on whether tag is set or not.